repo/content/index_blob_manager.go

```diff
@@ -0,0 +1,325 @@
+	}
+
+	return nil
+}
+
+func (m *indexBlobManagerImpl) getIndexBlob(ctx context.Context, blobID blob.ID) ([]byte, error) {
+	payload, err := m.indexBlobCache.getContent(ctx, cacheKey(blobID), blobID, 0, -1)
+	if err != nil {
+		return nil, err
+	}
+
+	iv, err := getIndexBlobIV(blobID)
+	if err != nil {
+		return nil, err
+	}
+
+	payload, err = m.encryptor.Decrypt(nil, payload, iv)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "decrypt error")
+	}
+
+	// Since the encryption key is a function of data, we must be able to generate exactly the same key
+	// after decrypting the content. This serves as a checksum.
+	if err := m.verifyChecksum(payload, iv); err != nil {
+		return nil, err
+	}
+
+	return payload, nil
+}
+
+func (m *indexBlobManagerImpl) verifyChecksum(data, contentID []byte) error {
+	var hashOutput [maxHashSize]byte
+
+	expected := m.hasher(hashOutput[:0], data)
+	expected = expected[len(expected)-aes.BlockSize:]
+
+	if !bytes.HasSuffix(contentID, expected) {
+		return errors.Errorf("invalid checksum for blob %x, expected %x", contentID, expected)
+	}
+
+	return nil
+}
+
+func (m *indexBlobManagerImpl) writeIndexBlob(ctx context.Context, data []byte) (blob.Metadata, error) {
+	return m.encryptAndWriteBlob(ctx, data, indexBlobPrefix)
+}
+
+func (m *indexBlobManagerImpl) encryptAndWriteBlob(ctx context.Context, data []byte, prefix blob.ID) (blob.Metadata, error) {
+	var hashOutput [maxHashSize]byte
+
+	hash := m.hasher(hashOutput[:0], data)
+	blobID := prefix + blob.ID(hex.EncodeToString(hash))
+
+	iv, err := getIndexBlobIV(blobID)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	data2, err := m.encryptor.Encrypt(nil, data, iv)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	m.listCache.deleteListCache(prefix)
+
+	err = m.st.PutBlob(ctx, blobID, gather.FromSlice(data2))
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	bm, err := m.st.GetMetadata(ctx, blobID)
+	if err != nil {
+		return blob.Metadata{}, errors.Wrap(err, "unable to get blob metadata")
+	}
+
+	if err := m.ownWritesCache.add(ctx, bm); err != nil {
+		log(ctx).Warningf("unable to cache own write: %v", err)
+	}
+
+	return bm, nil
+}
+
+func (m *indexBlobManagerImpl) getCompactionLogEntries(ctx context.Context, blobs []blob.Metadata) (map[blob.ID]*compactionLogEntry, error) {
+	results := map[blob.ID]*compactionLogEntry{}
+
+	for _, cb := range blobs {
+		data, err := m.getIndexBlob(ctx, cb.BlobID)
+
+		if errors.Is(err, blob.ErrBlobNotFound) {
+			continue
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &compactionLogEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (m *indexBlobManagerImpl) deleteOldIndexBlobs(ctx context.Context, latestBlob blob.Metadata) error {
```

Comment:
rename the function? Since we are also deleting compaction blobs, not just index blobs.

Reply:
done